redfish-core/lib/pcie.hpp

```diff
@@ -91,6 +92,235 @@ static inline void
+    {
+        resp.jsonValue["Status"]["State"] = "Unused";
+        resp.jsonValue["Status"]["Health"] = NULL;
+        return;
+    }
+
+    if (linkStatus ==
+        "xyz.openbmc_project.Inventory.Item.PCIeSlot.Status.Unknown")
+    {
+        resp.jsonValue["Status"]["State"] = NULL;
+        resp.jsonValue["Status"]["Health"] = NULL;
+    }
+}
+
+/**
+ * @brief Get PCIeSlot properties.
+ * @param[in,out]   asyncResp       Async HTTP response.
+ * @param[in]       pcieSlotPath    Object path of the PCIeSlot.
+ * @param[in]       serviceMap      A map to hold Service and corresponding
+ * interface list for the given cable id.
+ */
+static inline void
+    getPcieSlotLinkStatus(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                          const std::string& pcieSlotPath,
+                          const dbus::utility::MapperServiceMap& serviceMap)
+{
+    for (const auto& [service, interfaces] : serviceMap)
+    {
+        for (const auto& interface : interfaces)
+        {
+            if (interface != "xyz.openbmc_project.Inventory.Item.PCIeSlot")
+            {
+                continue;
+            }
+
+            crow::connections::systemBus->async_method_call(
+                [asyncResp](
+                    const boost::system::error_code ec,
+                    const dbus::utility::DBusPropertiesMap& properties) {
+                    if (ec)
+                    {
+                        BMCWEB_LOG_DEBUG << "DBUS response error " << ec;
+                        messages::internalError(asyncResp->res);
+                        return;
+                    }
+
+                    for (const auto& [propKey, propVariant] : properties)
+                    {
+
+                        if (propKey != "LinkStatus")
+                        {
+                            continue;
+                        }
+
+                        const std::string* value =
+                            std::get_if<std::string>(&propVariant);
+                        if (value == nullptr)
+                        {
+                            messages::internalError(asyncResp->res);
+                            return;
+                        }
+                        std::string linkStatus = *value;
+                        if (!linkStatus.empty())
+                        {
+                            fillPcieDeviceStatus(asyncResp->res, linkStatus);
+                            return;
+                        }
+                    }
+                    BMCWEB_LOG_ERROR << "PCIe Slot LinkStatus property not "
+                                        "found, treating it as 'Unknown'";
+                    fillPcieDeviceStatus(asyncResp->res,
+                                         "xyz.openbmc_project.Inventory.Item."
+                                         "PCIeSlot.Status.Unknown");
+                },
+                service, pcieSlotPath, "org.freedesktop.DBus.Properties",
+                "GetAll", interface);
```

Comment:
Do we need a GetAll call here on PCIeSlot interface or can we just make a call to LinkStatus? That way we can avoid looping through properties?

Reply:
Get call should be good here as we are only interested in single property. Good one!